framework/src/org/checkerframework/common/value/ValueTransfer.java

```diff
@@ -278,8 +380,50 @@ private void refineArrayAtLengthAccess(FieldAccessNode arrayLengthNode, CFStore
         if (!NodeUtils.isArrayLengthFieldAccess(arrayLengthNode)) {
             return;
         }
+
+        refineAtLengthAccess(arrayLengthNode, arrayLengthNode.getReceiver(), store);
+    }
+
+    /**
+     * If string.length() is encountered, transform its @IntVal annotation into an @ArrayLen
+     * annotation for string.
+     */
+    private void refineStringAtLengthInvocation(
+            MethodInvocationNode stringLengthNode, CFStore store) {
+        MethodAccessNode methodAccessNode = stringLengthNode.getTarget();
+
+        if (atypefactory.isStringLengthMethod(methodAccessNode.getMethod())) {
+            refineAtLengthAccess(stringLengthNode, methodAccessNode.getReceiver(), store);
+        }
+    }
+
+    /** Gets a value checker annotation relevant for an array or a string. */
+    private AnnotationMirror getArrayAnnotation(Node arrayNode) {
+        AnnotationMirror arrayAnno =
+                atypefactory.getAnnotationMirror(arrayNode.getTree(), StringVal.class);
+        if (arrayAnno == null) {
+            arrayAnno = atypefactory.getAnnotationMirror(arrayNode.getTree(), ArrayLen.class);
+        }
+        if (arrayAnno == null) {
+            arrayAnno = atypefactory.getAnnotationMirror(arrayNode.getTree(), ArrayLenRange.class);
+        }
+
+        return arrayAnno;
+    }
+
+    /**
+     * Transform @IntVal or @IntRange annotations of a array or string length into an @ArrayLen
+     * or @ArrayLenRange annotation for the array or string.
+     */
+    private void refineAtLengthAccess(Node lengthNode, Node receiverNode, CFStore store) {
         Receiver arrayLenRec =
-                FlowExpressions.internalReprOf(analysis.getTypeFactory(), arrayLengthNode);
+                FlowExpressions.internalReprOf(analysis.getTypeFactory(), lengthNode);
+
+        // If the expression is not representable (for example if String.length() for some reason is not marked @Pure, then do not refine.
+        if (arrayLenRec instanceof FlowExpressions.Unknown) {
```

Comment:
Probably should change the name of `arrayLenRec` so that it no longer mentions arrays. You changed some of these names but not all of them.